compiler/src/dotty/tools/dotc/util/CityHash.scala

```diff
@@ -0,0 +1,186 @@
+      hashLen16(c, d, mul)
+    } else if (len >= 4) {
+      val mul = k2 + len * 2
+      val a = fetch32(0)
+      hashLen16(len + (a << 3), fetch32(len - 4), mul)
+    } else if (len > 0) {
+      val a: Byte = data(0)
+      val b: Byte = data(len >> 1)
+      val c: Byte = data(len - 1)
+      val y: Int = a.toInt + (b.toInt << 8)
+      val z: Int = len + (c.toInt << 2)
+      shiftMix(y * k2 ^ z * k0) * k2
+    } else {
+      k2
+    }
+  }
+
+  // This probably works well for 16-byte strings as well, but it may be overkill
+  // in that case.
+  private final def hashLen17to32(len: Int)(implicit data: Array[Byte]): Long = {
+    val mul: Long = k2 + len * 2
+    val a: Long = fetch64(0) * k1
+    val b: Long = fetch64(8)
+    val c: Long = fetch64(len - 8) * mul
+    val d: Long = fetch64(len - 16) * k2
+    hashLen16(rotate(a + b, 43) + rotate(c, 30) + d, a + rotate(b + k2, 18) + c, mul)
+  }
+
+  /** Return an 8-byte hash for 33 to 64 bytes. */
+  private final def hashLen33to64(len: Int)(implicit data: Array[Byte]): Long = {
+    val mul: Long = k2 + len * 2
+    val a = fetch64(0) * k2
+    val b = fetch64(8)
+    val c = fetch64(len - 24)
+    val d = fetch64(len - 32)
+    val e = fetch64(16) * k2
+    val f = fetch64(24) * 9
+    val g = fetch64(len - 8)
+    val h = fetch64(len - 16) * mul
+    val u = rotate(a + g, 43) + (rotate(b, 30) + c) * 9
+    val v = ((a + g) ^ d) + f + 1
+    val w = bswap64((u + v) * mul) + h
+    val x = rotate(e + f, 42) + c
+    val y = (bswap64((v + w) * mul) + g) * mul
+    val z = e + f + c
+    val a2 = bswap64((x + z) * mul + y) + b
+    shiftMix((z + a2) * mul + d + h) * mul + x
+  }
+
+  private final def hashLen16(hi: Long, lo: Long): Long = {
+    val kMul = 0x9ddfea08eb382d69L
+    var a = (lo ^ hi) * kMul
+    a ^= (a >> 47)
+    var b = (hi ^ a) * kMul
+    b ^= (b >> 47)
+    b * kMul
+  }
+
+  private final def hashLen16(u: Long, v: Long, mul: Long): Long = {
+    // Murmur-inspired hashing.
+    var a = (u ^ v) * mul
+    a ^= (a >> 47)
+    a = (v ^ a) * mul
+    a ^= (a >> 47)
+    a * mul
+  }
+
+  /** Return a 16-byte hash for 48 bytes.  Quick and dirty.
+   *  Callers do best to use "random-looking" values for a and b.
+   */
+  private final def weakHashLen32WithSeeds(w: Long, x: Long, y: Long, z: Long, a0: Long, b0: Long): (Long, Long) = {
+    var a = a0
+    var b = b0
+    a += w
+    b = rotate(b + a + z, 21)
+    val c: Long = a
+    a += x
+    a += y
+    b += rotate(a, 44)
+    (a + z, b + c)
+  }
+
+  /** Return a 16-byte hash for s[0] ... s[31], a, and b.  Quick and dirty. */
+  private final def weakHashLen32WithSeeds(s: Int, a: Long, b: Long)(implicit data: Array[Byte]): (Long, Long) =
+    weakHashLen32WithSeeds(fetch64(s), fetch64(s + 8), fetch64(s + 16), fetch64(s + 24),  a, b)
+
+  private final def fetch64(idx: Int)(implicit data: Array[Byte]): Long = {
+    var x: Long = data(idx)
+    x = data(idx + 1) | (x << 8)
```

Comment:
Careful with sign extension (because `Byte` is signed). If I recall correctly, you need to add`& 0xff` everywhere you convert from a `Byte` to an `Int` to avoid that.